src/groth16/aggregate/verify.rs

```diff
@@ -0,0 +1,639 @@
+                // let agg_c = inner_product::multiexponentiation::<E::G1Affine>(&c, r_vec)
+                &proof.agg_c.into_affine().prepare(),
+                &pvk.delta_g2,
+            )]));
+            p3.send(tuple).unwrap();
+        });
+
+        let (r_vec_sender, r_vec_receiver) = bounded(1);
+        s.spawn(move |_| {
+            let now = Instant::now();
+            r_vec_sender
+                .send(structured_scalar_power(public_inputs.len(), &r))
+                .unwrap();
+            let elapsed = now.elapsed().as_millis();
+            debug!("generation of r vector: {}ms", elapsed);
+        });
+
+        // 5. compute the middle part of the final pairing equation, the one
+        //    with the public inputs
+        //let p2 = send_tuple.clone();
+        s.spawn(move |_| {
+            // We want to compute MUL(i:0 -> l) S_i ^ (SUM(j:0 -> n) ai,j * r^j)
+            // this table keeps tracks of incremental computation of each i-th
+            // exponent to later multiply with S_i
+            // The index of the table is i, which is an index of the public
+            // input element
+            // We incrementally build the r vector and the table
+            // NOTE: in this version it's not r^2j but simply r^j
+
+            let l = public_inputs[0].len();
+            let mut g_ic = pvk.ic_projective[0];
+            g_ic.mul_assign(r_sum);
+
+            let powers = r_vec_receiver.recv().unwrap();
+
+            let now = Instant::now();
+            // now we do the multi exponentiation
+            let getter = |i: usize| -> <E::Fr as PrimeField>::Repr {
+                // i denotes the column of the public input, and j denotes which public input
+                let mut c = public_inputs[0][i];
+                for j in 1..public_inputs.len() {
+                    let mut ai = public_inputs[j][i];
+                    ai.mul_assign(&powers[j]);
+                    c.add_assign(&ai);
+                }
+                c.into_repr()
+            };
+
+            let totsi = par_multiscalar::<_, E::G1Affine>(
+                &ScalarList::Getter(getter, l),
+                &pvk.multiscalar.at_point(1),
+                std::mem::size_of::<<E::Fr as PrimeField>::Repr>() * 8,
+            );
+
+            g_ic.add_assign(&totsi);
+
+            let tuple = PairingCheck::from_miller_one(E::miller_loop(&[(
+                &g_ic.into_affine().prepare(),
+                &pvk.gamma_g2,
+            )]));
+            let elapsed = now.elapsed().as_millis();
+            debug!("table generation: {}ms", elapsed);
+
+            send_tuple.send(tuple).unwrap();
+        });
+
+        s.spawn(move |_| {
+            // final value ip_ab is what we want to compare in the groth16
+            // aggregated equation A * B
+            let mut acc = PairingCheck::from_pair(E::Fqk::one(), proof.ip_ab.clone());
+            while let Ok(tuple) = rcv_tuple.recv() {
+                acc.merge(&tuple);
```

Comment:
I don't understand why not all of the pairs need to be randomized, looking at the batch verification there are random factors generated for each of the pairs, but here it seems only some are getting the randomization and some not

Reply:
I think we clarified this in Slack but the gist is any two pairs of pairing checks need to have at least one random component. so by randomizing all checks but only one, we can satisfy that property.